### PR TITLE
Add indeterminate progress bar below EngDiff FitPropertyBrowser

### LINKED ISSUE #34414: EngDiff Fitting Feedback
From email on 17/08/22

**Original reporter:** Saurabh

**Describe**
The suggestion is some kind of feedback to know when the fit is running and then when completed. I believe this was raised relating to the fitting tab but probably also relevant for the GSASII tab.

"You get almost no feedback so you don’t know what is going on and then everything appears but since the matrix workspaces are hidden within a list of other WS you don’t see it till you actually look for it.

### PR DESCRIPTION
**Pull request by DanielMurphy22**
**Description of work.**
Add new indeterminate progress bar to show when fitting is happening or has finished on the EngDiff fitting tab.

**To test:**

Please try to break this new progress bar in as many ways as possible! And together we can check that its style is clear on all OS (I've checked Ubuntu and MacOS so far). It has a moving partially blue bar when fitting, solid green for successful fit and empty with red border for failed fit. The progress bar resets to empty with a normal grey border when the fit browser is toggled open/close or a workspace is added or removed from the plot. The progress bar should mostly be in sync with the status reported text at the top of the embedded fit property browser.

- Load some focussed data into the EngDiff fitting tab and add it to the plot. Click the Fit toolbar button to enable the FitPropertyBrowser. Right-click in the plot and add a peak. 
- Click Fit > Fit and notice how the progress bar works.
- Add a few more peaks to get a longer fitting time.
- When a peaks are added to the model you should be able to click "Sequential Fit" or "Serial Fit" in the data table just above the plot. Check these methods of fitting also interact with the progress bar well.
- When sequentially or serially fitting it is possible that some spectra may succeed and some may fail. The progress bar should be in sync with the reported status at the top of the Fit property browser and report the status for the last fitted spectrum!
- again please try and break this as much as possible!
- check if tests are good enough and if I need more, I'd really like this new feature to be reliable for the users!

Fixes #34414 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by RichardWaiteSTFC** [comment]
On windows if you start a fit and then immediately delete the run from the table (not sure I managed to do this before the fit had finished but think i was quick enough to do it before the plot updated) you can get mantid to crash. But I don't know if this is also a problem on main.

**Comment by RichardWaiteSTFC** [comment]
> On windows if you start a fit and then immediately delete the run from the table (not sure I managed to do this before the fit had finished but think i was quick enough to do it before the plot updated) you can get mantid to crash. But I don't know if this is also a problem on main.

I can't reproduce this exactly, but following the above instructions results in this error (on this branch and main) - presumably because the plot clears itself and closes the fit browser if only one run is plotted.
```
Traceback (most recent call last):
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\widgets\fitpropertybrowser\fitpropertybrowser.py", line 485, in peak_changed_slot
 self.tool.update_peak(peak_id, self.getPeakCentreOf(prefix),
AttributeError: 'NoneType' object has no attribute 'update_peak'
```

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
Works well, there's a slight difference in behaviour between serial/sequential fits and normal fits - the former lock the interface 
https://github.com/mantidproject/mantid/blob/b9b58ef761e260f5b94085d208586435572a4355/qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/fitting/presenter.py#L18-L19
which I think would be good to do in the case of the normal fit using the new `fit_started/complete_observer` you added - what do you think?

Another minor thing is that for serial/sequential fits it looks like the progress bar is updated with the final fit status before the workspaces with the fit byproducts are grouped, whereas when calling `Fit` from the fit property browser there is a  lag where the fit has finished according to the log but the workspaces are being made.  Other than that it works well, the status is cleared on adding workspaces, closing the fit property browser etc. which I think is the correct behaviour. And the tests look good, thanks.

**Comment by RichardWaiteSTFC** [comment]
FYI I have come a couple of unrelated bugs - have you seen this/similar before?
1. Found this when performing a serial fit which presumably did not converge on something sensible - it looks there should be a check that all the fit parameters are finite before passing proceeding to the next run and passing them to the `Fit` algorithm?
```
Error in execution of algorithm Fit:
Trying to set a NaN value (-nan(ind)) to parameter I
Traceback (most recent call last):
 File "c:\mantid-conda\mantid\qt\python\mantidqtinterfaces\mantidqtinterfaces\Engineering\gui\engineering_diffraction\tabs\fitting\data_handling\data_presenter.py", line 62, in _start_serial_fit
 self.fit_all_started_notifier.notify_subscribers(ws_name_list, do_sequential=False)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\observer_pattern.py", line 99, in notify_subscribers
 Observable.notify_subscribers(self, *args)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\observer_pattern.py", line 55, in notify_subscribers
 self.thread_safe_update_call(arg, **kwargs)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\qt\qappthreadcall.py", line 67, in __call__
 return self._callable(*args)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\observer_pattern.py", line 59, in _notify_subscribers_impl
 observer.update(self, arg, **kwargs)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\observer_pattern.py", line 91, in update
 self.callback(arg)
 File "c:\mantid-conda\mantid\qt\python\mantidqtinterfaces\mantidqtinterfaces\Engineering\gui\engineering_diffraction\tabs\fitting\plotting\plot_presenter.py", line 80, in do_fit_all
 fit_output = Fit(**fitprop['properties'])
 File "C:\mantid-conda\mantid\Framework\PythonInterface\mantid\simpleapi.py", line 313, in wrapper
 algm.execute()
ValueError: Trying to set a NaN value (-nan(ind)) to parameter I
```
2. Not sure exactly how I produced this, I think it was deleting a run and the re-loading it and then performing a serial fit
```
Traceback (most recent call last):
 File "c:\mantid-conda\mantid\qt\python\mantidqtinterfaces\mantidqtinterfaces\Engineering\gui\engineering_diffraction\tabs\fitting\data_handling\data_presenter.py", line 62, in _start_serial_fit
 self.fit_all_started_notifier.notify_subscribers(ws_name_list, do_sequential=False)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\observer_pattern.py", line 99, in notify_subscribers
 Observable.notify_subscribers(self, *args)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\observer_pattern.py", line 55, in notify_subscribers
 self.thread_safe_update_call(arg, **kwargs)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\qt\qappthreadcall.py", line 67, in __call__
 return self._callable(*args)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\observer_pattern.py", line 59, in _notify_subscribers_impl
 observer.update(self, arg, **kwargs)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\observer_pattern.py", line 91, in update
 self.callback(arg)
 File "c:\mantid-conda\mantid\qt\python\mantidqtinterfaces\mantidqtinterfaces\Engineering\gui\engineering_diffraction\tabs\fitting\plotting\plot_presenter.py", line 96, in do_fit_all
 self.fit_all_done_notifier.notify_subscribers(fitprop_list)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\observer_pattern.py", line 99, in notify_subscribers
 Observable.notify_subscribers(self, *args)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\observer_pattern.py", line 55, in notify_subscribers
 self.thread_safe_update_call(arg, **kwargs)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\qt\qappthreadcall.py", line 67, in __call__
 return self._callable(*args)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\observer_pattern.py", line 59, in _notify_subscribers_impl
 observer.update(self, arg, **kwargs)
 File "c:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\utils\observer_pattern.py", line 91, in update
 self.callback(arg)
 File "c:\mantid-conda\mantid\qt\python\mantidqtinterfaces\mantidqtinterfaces\Engineering\gui\engineering_diffraction\tabs\fitting\data_handling\data_presenter.py", line 54, in fit_completed
 self.model.update_fit(fit_props)
 File "c:\mantid-conda\mantid\qt\python\mantidqtinterfaces\mantidqtinterfaces\Engineering\gui\engineering_diffraction\tabs\fitting\data_handling\data_model.py", line 348, in update_fit
 key = '_'.join([fname, params_dict['Name'][irow].split('.')[-1]]) # funcname_param
IndexError: list index out of range
```

File: instrument/sampleenvironments/ISIS/FlatPlateFrame.xml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+<!-- Definition of FlatPlate-->
```
Comment (RichardWaiteSTFC) [inline instrument/sampleenvironments/ISIS/FlatPlateFrame.xml:1]: Rest of the code looks good (in particular running the fit all stuff in an async task and tidying up the observers and putting them together in the fitting presenter) - just this one file that I think wasn't meant to be included?
About to give it a test!

Comment (DanielMurphy22) [reply instrument/sampleenvironments/ISIS/FlatPlateFrame.xml]: I've removed it now, great spot!

**Review by RichardWaiteSTFC — APPROVED** [review]
Works well, fits using the fit browser now disable the interface which makes it more difficult for the users to break things! The new cancel button for  serial and sequential fit works well and doesn't trigger the fit tables to be remade. Thanks for the changes!

**Pull request merged**